src/charm.py

```diff
@@ -6,20 +6,355 @@
+                )
+            except jinja2.UndefinedError as e:
+                undefined_key = e.message.split()[0].strip(r"'")
+                raise TemplateKeyError(obj, undefined_key) from e
+
+        url = _render(self.root_url)
+        if not self.rule:
+            rule = self.generate_rule_from_url(url)
+        else:
+            rule = _render(self.rule)
+
+        # an easily recognizable id for the traefik services
+        id_ = "-".join((unit_name, model_name))
+        return RouteConfig(rule=rule, root_url=url, id_=id_)
+
+    @staticmethod
+    def generate_rule_from_url(url: str) -> Optional[str]:
+        """Derives a Traefik router Host rule from the provided `url`'s hostname."""
+        url_ = urlparse(url)
+        if not url_.hostname:
+            logger.info(f"Parsing url {url_} not valid")
+            raise RuleDerivationError(url)
+        return f"Host(`{url_.hostname}`)"
+
+
+def _is_not_empty(config: str) -> bool:
+    return bool(config and not config.isspace())
+
+
 class OathkeeperConfiguratorCharm(CharmBase):
     """Charmed Oathkeeper Configurator."""
 
-    def __init__(self, *args):
+    _stored = StoredState()
+    _ingress_relation_name = "ingress-per-unit"
+
+    def __init__(self, *args: Any) -> None:
         super().__init__(*args)
 
+        self._stored.set_default(invalid_config=False)
+
+        self.ingress_per_unit = IngressPerUnitProvider(self, self._ingress_relation_name)
+
+        # Charm events
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+        # Relation events
+        self.framework.observe(
+            self.ingress_per_unit.on.data_provided, self._on_ingress_data_provided
+        )
+
+    @property
+    def access_rules_configured(self) -> bool:
+        """Checks whether the access rules have been configured."""
+        return _is_not_empty(self.model.config.get("access_rules", ""))
+
+    def validate_access_rules_config(self, config: str) -> None:
+        """Validate that the access rules config is a valid json."""
+        try:
+            json.loads(config)
+        except json.JSONDecodeError as e:
+            raise InvalidConfigError(f"Failed to decode json: {e}")
+
+    @property
+    def _config(self) -> _RouteConfig:
+        return _RouteConfig(rule=self.config.get("rule"), root_url=self.config.get("root_url"))
+
+    @property
+    def rule(self) -> Optional[str]:
+        """The Traefik rule this charm is responsible for configuring."""
+        return self._config.rule
+
+    @property
+    def root_url(self) -> Optional[str]:
+        """The advertised url for the charm requesting ingress."""
+        return self._config.root_url
+
+    def _render_config(self, model_name: str, unit_name: str, app_name: str) -> RouteConfig:
+        return self._config.render(model_name=model_name, unit_name=unit_name, app_name=app_name)
+
+    @staticmethod
+    def _generate_traefik_unit_config(config: RouteConfig) -> UnitConfig:
```

Comment:
It looks like the logic about traefik unit config generation, merging, checking if it is ready, etc. is nothing to do with the charm itself. Do you think it is better we extract them into an individual class?

Reply:
I could extract some of that logic into an individual class, but most of it is coupled with the charm, e.g. when it relies on charm relations, config or unit status, or uses charm's stored state to store info on config's validity.